File: src/lib/services/check-endpoint.ts
import type { IEndpoint } from "@/lib/models/endpoint";
import type { CheckResultData } from "@/types";

export async function checkEndpoint(
  endpoint: IEndpoint
): Promise<CheckResultData> {
  const controller = new AbortController();
  const timeoutId = setTimeout(
    () => controller.abort(),
    endpoint.timeout * 1000
  );

  const start = Date.now();
  try {
    const headers: Record<string, string> = {};
    if (endpoint.headers) {
      endpoint.headers.forEach((value: string, key: string) => {
        headers[key] = value;
      });
    }

    const res = await fetch(endpoint.url, {
      method: endpoint.method,
      headers,
      body: ["POST", "PUT", "PATCH"].includes(endpoint.method)
        ? endpoint.body || undefined
        : undefined,
      signal: controller.signal,
    });

    clearTimeout(timeoutId);
    const responseTime = Date.now() - start;

    let status: CheckResultData["status"];
    if (res.status === endpoint.expectedStatusCode) {
      status = "UP";
    } else if (res.status >= 500) {
      status = "DOWN";
    } else {
      status = "DEGRADED";
    }

    return {
      status,
      statusCode: res.status,
      responseTime,
      error: null,
    };
  } catch (err) {
    clearTimeout(timeoutId);
    const responseTime = Date.now() - start;
    return {
      status: "DOWN",
      statusCode: null,
      responseTime,
      error: (err as Error).message,
    };
  }
}

File: backend/services/check_endpoint.py
import httpx

from models.endpoint import Endpoint
from services.check_db import PROBES as _DB_PROBES


async def check_endpoint(endpoint: Endpoint) -> dict:
    # DB endpoints route to protocol-specific probes. Same return shape
    # as the HTTP path, so notification + alert flow stays uniform.
    kind = getattr(endpoint, "kind", "http") or "http"
    if kind in _DB_PROBES:
        return await _DB_PROBES[kind](endpoint)

    start = 0
    try:
        headers = dict(endpoint.headers) if endpoint.headers else {}
        body_content = endpoint.body if endpoint.method in ("POST", "PUT", "PATCH") and endpoint.body else None

        async with httpx.AsyncClient(timeout=endpoint.timeout) as client:
            import time
            start = time.monotonic()
            resp = await client.request(
                method=endpoint.method,
                url=endpoint.url,
                headers=headers,
                content=body_content,
            )
            response_time = round((time.monotonic() - start) * 1000, 2)

        if resp.status_code == endpoint.expected_status_code:
            status = "UP"
        elif resp.status_code >= 500:
            status = "DOWN"
        else:
            status = "DEGRADED"

        return {
            "status": status,
            "status_code": resp.status_code,
            "response_time": response_time,
            "error": None,
        }
    except Exception as exc:
        import time
        response_time = round((time.monotonic() - start) * 1000, 2) if start else 0
        return {
            "status": "DOWN",
            "status_code": None,
            "response_time": response_time,
            "error": str(exc),
        }
